Cm/fixes for big o (#64)
Ada-Developers-Academy/core-unit-1: big-o/big-o-notation.md

| O(n<sup>2</sup>) | Quadratic   | The algorithm will have a runtime or memory usage proportional to the size of the input squared. This often involves 2 nested loops.                                                                                                                                                    |
| O(2<sup>n</sup>) | Exponential | The algorithm's complexity doubles each time the input size increases by one.                                                                                                                                      |

You can see a chart comparing these complexities below.  This chart illustrates for a given input size, how the output of the given expressions change.  Look at the curve of each line.  Which curve accellerates most quickly?  Which curve barely increases at all as the input size increases.

> ![Complexities comparison chart](images/time-complexities.png)

Reviewer: Chris hehehehe sorry to be picky about files (actually, you know I'm not sorry, but I'm still sorry)

Could you throw all assets into `../assets` (aka there's an `assets` folder on the root level)? Alooosososososo my preferred fiiile naming pattern for this would be `big-o-notation_time-complexities.png` loooool

Also, weirdly enough, this picture doesn't show up in my GH preview anyway. Is that a problem on my preview end?

Author: I've thrown it into assets and renamed it.

I think the preview must be on your end as it shows up for me and in the branch on github.

### !callout-info

## How do we say it out loud?

Developers talk about Big O a lot, so it's useful to know how to talk about this out loud. Here are some example sentences to help start the pattern:

> - "The Big O of linear search is O of n. It runs in linear time."

Reviewer: Wow I really just mixed up my constants and linears in this one. hehehe thanks

- "An algorithm with a Big O of log n is better than O of n log n."
- "O of n-squared is one of the slowest!"
- "The algorithm does a lot of work on each item in the input, but it's still order n!"

### !end-callout

<!-- available callout types: info, success, warning, danger, secondary  -->
### !callout-warning

> ## Big-O Doesn't give an exact number

Reviewer: Rest of the lesson doesn't use "Big-O," so conforming to one style

```suggestion
## Big O Doesn't give an exact number
```

> ## Big O Doesn't give an exact number

Reviewer 2: casing
```suggestion
## Big O Doesn't Give an Exact Number
```

<!-- Question about comparing Big Os for more efficient -->

<!-- >>>>>>>>>>>>>>>>>>>>>> BEGIN CHALLENGE >>>>>>>>>>>>>>>>>>>>>> -->
<!-- Replace everything in square brackets [] and remove brackets  -->

### !challenge

* type: ordering
* id: d2d156f6-aaab-4aba-9cf7-b9cdaceed013
* title: Time/Space Complexities Worst-to-best
* points: 1
* topics: big-o

##### !question

> Arrange the following complexities from worst to best

Reviewer: ```suggestion
Arrange the following complexities from worst to best.
```

```suggestion
Arrange the following complexities from least efficient to most efficient.
```

##### !end-question

##### !answer

1. O(1)
1. O(2^n)
1. O(n log n)
1. O(log n)
1. O(n^2)
1. O(n)

##### !end-answer

<!-- other optional sections -->
<!-- !hint - !end-hint (markdown, hidden, students click to view) -->
<!-- !rubric - !end-rubric (markdown, instructors can see while scoring a checkpoint) -->
<!-- !explanation - !end-explanation (markdown, students can see after answering correctly) -->

### !end-challenge

<!-- ======================= END CHALLENGE ======================= -->

<!-- >>>>>>>>>>>>>>>>>>>>>> BEGIN CHALLENGE >>>>>>>>>>>>>>>>>>>>>> -->
<!-- Replace everything in square brackets [] and remove brackets  -->

### !challenge

* type: multiple-choice
* id: db5bbef4-04da-4474-8aa1-5a86e7047e6a
* title: Purpose of Big-O Notation
* points: 1
* topics: Big-o

##### !question

> Big-O notation serve to:

Reviewer: - Conform "Big O" style
- "serve" to "serves"

```suggestion
Big O notation serves to:
```

* Describe only how well an algorithm will perform in terms of **speed** as the input size increases
* Describe only how well an algorithm will perform in terms of **memory usage** as the input size increases
  
##### !end-options

##### !answer

* Describe how the algorithm performs in terms of both space and time as the input size increases

##### !end-answer

<!-- other optional sections -->
<!-- !hint - !end-hint (markdown, hidden, students click to view) -->
<!-- !rubric - !end-rubric (markdown, instructors can see while scoring a checkpoint) -->
<!-- !explanation - !end-explanation (markdown, students can see after answering correctly) -->

### !end-challenge

<!-- ======================= END CHALLENGE ======================= -->

<!-- Question about comparing Big Os for more efficient -->

<!-- >>>>>>>>>>>>>>>>>>>>>> BEGIN CHALLENGE >>>>>>>>>>>>>>>>>>>>>> -->
<!-- Replace everything in square brackets [] and remove brackets  -->

### !challenge

* type: multiple-choice
* id: 6b2620b0-f164-4722-8a55-a000a6b3690a
* title: Pick the best performing algorithm
* points: 1
* topics: Big-O

##### !question

> Given the following algorithms for a contact list project, which is the best choice in terms of time/space complexity?

Reviewer: ```suggestion
Imagine creating an app that allows a user to store all of their contacts and contact information. There is a feature that allows users to search for a specific contact.

Given the following algorithms for this search feature, which is the best choice in terms of time and space complexity?
```

##### !end-question

##### !options

* Store all contacts in a list and seach for each contact, iterating from beginning until end until the contact is found or the list is completely searched.  O(n)
* Store all contacts in a dictionary with the contact name as a key and retrieve contacts using the name.  O(1)

> * Store all contacts in a list sorted by names and search for a contact by looking at the middle contact.  If that is the desired contact return them.  If the desired contact comes later repeat the search in the 2nd half of the list, otherwise repeat the search in the 1st half of the remaining contacts.  Repeat until the contact is found or there are no more contacts left to search O(log n).

Reviewer: Conforming punctuation style to other options.

```suggestion
* Store all contacts in a list sorted by names and search for a contact by looking at the middle contact.  If that is the desired contact return them.  If the desired contact comes later repeat the search in the 2nd half of the list, otherwise repeat the search in the 1st half of the remaining contacts.  Repeat until the contact is found or there are no more contacts left to search. O(log n).
```

<!-- Question about comparing Big Os for more efficient -->

<!-- >>>>>>>>>>>>>>>>>>>>>> BEGIN CHALLENGE >>>>>>>>>>>>>>>>>>>>>> -->
<!-- Replace everything in square brackets [] and remove brackets  -->

### !challenge

* type: ordering
* id: d2d156f6-aaab-4aba-9cf7-b9cdaceed013
* title: Time/Space Complexities Worst-to-best
* points: 1
* topics: big-o

##### !question

Arrange the following complexities from **worst** to **best**.

##### !end-question

##### !answer

1. O(2^n)

> 1. O(n^2)

Reviewer 2: Just one more thought that when these were introduced, you used the nice superscript stuff.  Maybe do that here?  Or introduce the more computer-y style of writing it above as well?

Author: Ok I've switched both to the `<sub>` notation.  I think the original lesson here came from two different sources which is why the discrepancy.

1. O(n log n)
1. O(log n)
1. O(n^2)

> 1. O(n)

Reviewer 2: How does Learn know the order for this?  Is it supposed to be listed in the correct order and Learn will scramble it?  If so, I'm confused, since the problem says to list these from worst to best, but at either extremity we have O(1) and O(n), which are both pretty dang good...

Am I missing something?

Reviewer: It should be listed in the correct order, Ansel's totally correct!
